EIPS/eip-5164.md

```diff
@@ -0,0 +1,216 @@
+
+## Motivation
+
+Many Ethereum protocols need to coordinate state changes across multiple EVM-based blockchains. These chains often have native or third-party bridges that allow Ethereum contracts to execute code. However, bridges have different APIs so bridge integrations are bespoke.
+
+Bridge technology is improving rapidly, and there are multiple bridges to choose from. Each one affords different properties; with varying degrees of security, speed, and control.
+
+By standardizing a cross-chain execution interface, we can cleanly separate the transport layer from the application layer. Cross-chain execution becomes another composable piece with which the Ethereum ecosystem can build shared infrastructure.
+
+## Specification
+
+This specification allows contracts on one chain to send messages to contracts on another chain. There are two key contracts:
+
+- CrossChainRelayer
+- CrossChainReceiver
+
+The `CrossChainRelayer` lives on the origin chain. Messages sent to this contract are relayed to CrossChainReceivers on other chains.
+
+The `CrossChainReceiver` lives on the destination chain. The receiver is bound to a CrossChainRelayer and executes any messages sent by the relayer.
+
+When a user wishes to send cross-chain messages, they will create a CrossChainRelayer and a CrossChainReceiver.
+
+The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in [RFC 2119](https://datatracker.ietf.org/doc/html/rfc2119).
+
+### CrossChainRelayer
+
+The `CrossChainRelayer` lives on the chain from which messages are sent. The Relayer's job is to broadcast the messages through a transport layer.
+
+#### Methods
+
+**relayCalls**
+
+MUST emit the `Relayed` event when successfully called.
+
+MUST increment a `nonce` so that each batch of calls can be uniquely identified.
+
+```solidity
+struct Call {
+  address target;
+  bytes data;
+  uint value;
+}
+
+interface CrossChainRelayer {
+    function relayCalls(Call[] calldata calls) external;
+}
+```
+
+```yaml
+- name: relayCalls
+  type: function
+  stateMutability: nonpayable
+  inputs: 
+    - name: calls
+      type: Call[]
+```
+
+#### Events
+
+**Relayed**
+
+MUST be emitted when a CrossChainRelayer relays calls.
+
+```solidity
+interface CrossChainRelayer {
+  event Relayed(
+      uint256 indexed nonce,
+      Call[] calls
+  );
+}
+```
+
+```yaml
+- name: Relayed
+  type: event
+  inputs:
+    - name: nonce
+      indexed: true
+      type: uint256
+    - name: calls
+      type: Call[]
+```
+
+### CrossChainReceiver
+
+The `CrossChainReceiver` contract executes messages sent by a `CrossChainRelayer`. The CrossChainReceiver must execute calls that are relayed by the bound CrossChainRelayer.
+
+#### Methods
+
+**relayer**
+
+Returns the address of the relayer that this receiver is bound to.
```

Comment:
Above you capitalized Relayer, and here it's lowercase. Please choose a consistent capitalization throughout the document.